micro_sam/sam_annotator/annotator_2d.py

```diff
@@ -224,39 +285,51 @@ def annotator_2d(
     global PREDICTOR, IMAGE_EMBEDDINGS, AMG
     AMG = None
 
+    if raw is None:
+        raise RuntimeError("You must provide a raw input image source.")
+
+    # Check if user provided non-zero input values to tile_shape and halo parameters
+    if all(tile_shape) is False:
+        tile_shape = None
+    if all(halo) is False:
+        halo = None
+
     if predictor is None:
-        PREDICTOR = util.get_sam_model(model_type=model_type)
+        PREDICTOR = util.get_sam_model(model_type=model_type.name)
     else:
         PREDICTOR = predictor
 
+    # TODO: check if a pre-computed image embedding already matching the raw input image exists
+    # and if so use it without re-computing (warn the user you are re-using a previously generated embedding).
+
+    # TODO: dispatch long running computations to napari thread_worker
+    # and connect return to _start_2d_annotation function
+    # TODO: also consider a progress bar for long running computations
+    # see https://github.com/pyapp-kit/magicgui/issues/577#issuecomment-1705555387
+    print("Computing image embedding...")
     IMAGE_EMBEDDINGS = util.precompute_image_embeddings(
         PREDICTOR, raw, save_path=embedding_path, ndim=2, tile_shape=tile_shape, halo=halo,
         wrong_file_callback=show_wrong_file_warning
     )
     if precompute_amg_state and (embedding_path is not None):
+        print("Computing AMG...")
         AMG = cache_amg_state(PREDICTOR, raw, IMAGE_EMBEDDINGS, embedding_path)
 
+    # show the PCA of the image embeddings
+    if show_embeddings:
+        print("Show embeddings")
+        embedding_vis, scale = project_embeddings_for_visualization(IMAGE_EMBEDDINGS)
+        v.add_image(embedding_vis, name="embeddings", scale=scale)
+
     # we set the pre-computed image embeddings if we don't use tiling
     # (if we use tiling we cannot directly set it because the tile will be chosen dynamically)
     if tile_shape is None:
+        print("tiling precompute")
         util.set_precomputed(PREDICTOR, IMAGE_EMBEDDINGS)
 
-    # viewer is freshly initialized
-    if v is None:
```

Comment:
Just fyi: this condition is for the `image_series_annotator`: in that case we pass the napari viewer object via the `v` argument. Hence it is not `None` and `_update_viewer` is called to replace all relevant layers.
I just wanted to clarify because it relates to my comment on `_update_viewer`. We can go ahead and remove the condition since the `image_series_annotator` needs to be redone after finalizing the 2d annotator.